stdlib/src/memory/arc.mojo

```diff
@@ -45,32 +45,231 @@ print(Arc(String("ok"))[])
+                if cur_count == 0:
+                    return False
+
+                var success = self.refcount.compare_exchange_weak(
+                    cur_count, cur_count + 1
+                )
+
+                # TODO: should this be marked `unlikely` once we have intrinsic for it?
+                # This only occurs during races on `refcount` when the strong refs
+                # have churn.
+                if not success:
+                    continue
+
+                return True
+        else:
+            return False
 
     fn add_ref(inout self):
         """Atomically increment the refcount."""
         _ = self.refcount.fetch_add(1)
 
+        # any strong is also, underneath, a weak
+        @parameter
+        if enable_weak:
+            self.add_weak()
+
+    fn add_weak(inout self):
+        """Atomically increment the weakref count"""
+        @parameter
+        if enable_weak:
+            _ = self._weakref_ptr()[].fetch_add(1)
+
     fn drop_ref(inout self) -> Bool:
-        """Atomically decrement the refcount and return true if the result
-        hits zero."""
-        return self.refcount.fetch_sub(1) == 1
+        """
+        Atomically decrement the refcount and weakref count.
+        Returns:
+            True iff the backing allocation for `self`
+            should (must!) be deleted
+        """
+        var last_strong = self.refcount.fetch_sub(1) == 1
+
+        if last_strong:
+            self._destroy_value()
+
+        # defer to drop_weak() for whether the backing
+        # mem should be dealloc'd
+        @parameter
+        if enable_weak:
+            return self.drop_weak()
+        else:
+            return last_strong
+
+    fn drop_weak(inout self) -> Bool:
+        """
+        Atomically decrement the weakref count, and return
+        true if the result hits zero
+        Returns:
+            True iff the backing allocation for `self`
+            should (must!) be deleted
+        """
+        @parameter
+        if enable_weak:
+            var last_any = self._weakref_ptr()[].fetch_sub(1) == 1
+
+            # if this is the last weak, and there
+            # are no other strongs, then we can dealloc the inner
+            return last_any
+        else:
+            #TODO: statically ensure unreachable?
+            return False
+
+
+@register_passable
+struct Weak[T: Movable, enable_weak: Bool = True](CollectionElement, CollectionElementNew):
+    """A Weak[T] is used much like an Arc[T], except it does not
+    prevent the target instance of T from being deinitialized/deallocated.
+
+    This is useful for data structures like cyclic graphs and doubly linked lists,
+    such that users can easily avoid memory leaks.
+
+    Parameters:
+        T: The type of the value that this Weak may be upgraded into an Arc[T] of.
+    """
+
+    alias _inner_type = _ArcInner[T]
+    var _inner: UnsafePointer[Self._inner_type]
+
+    fn __init__(inout self, strong: Arc[T, enable_weak]):
```

Comment:
**Suggestion** Likewise, we can change this initializer so that `Weak[T]` can only be constructed frfom an `Arc[T, True]`:

```preview
          fn __init__(inout self, strong: Arc[T, True]):
```

This would let you remove the `constrained[enable_weak, "..."]()` below as well 🙂